### PR TITLE
Fix master builds to assess security concerns

### PR DESCRIPTION
**Pull request by jeancochrane**
# Overview

Builds on master were failing for a couple of reasons:

#### 1. A recent `gpg` update causes `gpg` to fail when run in an environment without a TTY([Debian bug report](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=913614#27))

To work around this bug, update the Slim Dockerfile to run GPG in batch mode as recommended by the [Docker Postgres maintainers](https://github.com/docker-library/postgres/issues/526#issuecomment-438818847). Batch mode instructs GPG to run non-interactively, as discussed in the Debian bug report:

> For any script use you should anyway use --batch which disables the use
of the tty as a side-effect.

> PS i do encourage everyone who is automating the use of gpg to use
--batch everywhere, as this forces GnuPG into a mode that is expected to
be used for automation (its "API", for lack of a better term, as opposed
to its "UI", which is its normal non-batch mode).
In addition to routing around the bug, this is also a more hygenic way of running GPG in CI.

#### 2. The version of `gdal` on the Alpine edge main repo has dependency resolution problems with `libcrypto` that prevents it from installing on Alpine builds

The latest version of `gdal` depends on `libcrypto 1.1`, but `libcrypto 1.1` is not available on the `v3.8/main` branch of Alpine, and is only available on `edge/main`. Even worse, it seems that `gdal` no longer exists in any Alpine repositories other than `edge/testing`.

To get around these dependency problems, instruct `apk` to look in both `edge/main` and `edge/testing` when resolving packages for `gdal`.
 
Some more context [in this issue](https://github.com/appropriate/docker-postgis/issues/96).

This PR fixes those issues in the following ways:

2. Bump Postgres minor versions for Stretch builds.
3. Install `gdal` on Alpine from the edge testing repo, as recommended here: https://github.com/appropriate/docker-postgis/issues/96

Connects https://github.com/azavea/american-water-tank-management/issues/84

## Notes

- This PR also updates `gevent` and `psycopg2` to the latest patch versions.
- This PR bumps the minor version of Postgres for Slim builds, introducing a minor version mismatch between the Postgres client on the two builds (Alpine runs `9.6.19` and `10.5` vs. Stretch's `9.6.20` and `10.6`). This seems reasonable to me since the [Postgres versioning policy](https://www.postgresql.org/support/versioning/) indicates that minor version upgrades shouldn't introduce breaking changes, and that upgrading minor versions should be considered safer than not.

## Testing

See https://travis-ci.org/azavea/docker-django/builds/460881083

### CODE REVIEW

**Comment by rbreslow** [comment]
@jeancochrane Which builds are failing? 

A quick search found this: https://github.com/docker-library/postgres/issues/526. 

I think it'd be more valuable to open an issue for this so we can track it and allocate time in the sprint to working on this.

**Comment by jeancochrane** [comment]
@rbreslow All of them, as you can see in https://travis-ci.org/azavea/docker-django/builds/442349816. This is the first step to addressing https://github.com/azavea/american-water-tank-management/issues/84.

**Comment by rbreslow** [comment]
Got it. I would look at that issue I referenced and apply the same fix as the official Docker hub images (https://github.com/docker-library/postgres/pull/527/files), and make sure to reference that discussion / the issue on the debian bug tracker in your PR.

**Comment by jeancochrane** [comment]
Good find! Apologies if this pulled your attention from something else, I meant to open it as a stub so I could run builds on Travis but I didn't realize it would ping everyone in Slack.

File: .travis.yml
Comment (rbreslow) [inline .travis.yml:5]: This is good.

As part of performing regular maintenance on this repository, we also check to see if there are any patch releases to dependencies within `X.Y/requirements.txt`.

I would reference existing PRs to see what normally changes during regular maintenance, and how we format the branch / PR body:

https://github.com/azavea/docker-django/pull/63

File: Dockerfile-slim.template
Comment (rbreslow) [inline Dockerfile-slim.template]: This is good. 

I think it would be valuable to include why the maintainers of the Docker Hub PostgreSQL container image are using `--batch` as a fix, and why it works.

If you see [this issue](https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=913614#27) on the Debian bug tracker, they state:

> For any script use you should anyway use --batch which disables the use
of the tty as a side-effect.

>PS i do encourage everyone who is automating the use of gpg to use
--batch everywhere, as this forces GnuPG into a mode that is expected to
be used for automation (its "API", for lack of a better term, as opposed
to its "UI", which is its normal non-batch mode).  And, FWIW, i agree
with Tianon that GnuPG should simply assume --batch if no tty exists,
but that's not the kind of change i can fit into debian stable, i think.

For future reference, Tianon is one of the lead maintainers of https://github.com/docker-library.

File: Dockerfile-slim.template
Comment (rbreslow) [inline Dockerfile-slim.template]: Clean up from #64, we should apply this diff, as we're not going to use the original solution from #56:

```diff
diff --git a/Dockerfile-slim.template b/Dockerfile-slim.template
index 82b0ae9..423b752 100644
--- a/Dockerfile-slim.template
+++ b/Dockerfile-slim.template
@@ -18,8 +18,7 @@ RUN set -ex; \
 # uid                  PostgreSQL Debian Repository
        key='B97B0AFCAA1A47F044F244A07FCC7D46ACCC4CF8'; \
        export GNUPGHOME="$(mktemp -d)"; \
-       (gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$key" \
-           || gpg --keyserver ipv4.pool.sks-keyservers.net --recv-keys "$key"); \
+       gpg --keyserver ha.pool.sks-keyservers.net --recv-keys "$key"; \
        gpg --export "$key" > /etc/apt/trusted.gpg.d/postgres.gpg; \
        rm -rf "$GNUPGHOME"; \
        apt-key list
```

File: Dockerfile-alpine.template
Comment (rbreslow) [inline Dockerfile-alpine.template:18]: It would be valuable to surface in the PR body what's going on here:

It looks like the version of `gdal` in the Alpine Linux package repositories used to depend on `libressl2.7-libcrypto`, which is available in the `v3.8/main` branch.

The default set of repositories configured in the root container image are:

```bash
/ # cat /etc/apk/repositories
http://dl-cdn.alpinelinux.org/alpine/v3.8/main
http://dl-cdn.alpinelinux.org/alpine/v3.8/community
```

If we want to look for packages from another repository, that repository needs to be explicitly stated with `--repository`.

The `gdal` package was updated to depend on `libcrypto1.1`, which is only available in the `edge/main` branch.

For more information on Alpine Linux package management, see: https://wiki.alpinelinux.org/wiki/Alpine_Linux_package_management

Also, small formatting nitpick:

```diff
diff --git a/Dockerfile-alpine.template b/Dockerfile-alpine.template
index 0f5e3dc..93615fe 100644
--- a/Dockerfile-alpine.template
+++ b/Dockerfile-alpine.template
@@ -16,8 +16,7 @@ RUN set -ex \
        gettext \
        postgresql-client=$PG_VERSION \
     && apk add --no-cache --repository http://dl-cdn.alpinelinux.org/alpine/edge/main \
-        libressl2.7-libcrypto \
-    && apk add --no-cache --repository http://dl-cdn.alpinelinux.org/alpine/edge/testing \
+       --repository http://dl-cdn.alpinelinux.org/alpine/edge/testing \
        gdal \
     && pip install --no-cache-dir -r /tmp/requirements.txt \
     && apk del .build-deps \
```

**Review by rbreslow — APPROVED** [review]
Good job digging through a rabbit hole, writing up a solid PR, and implementing a fix. 

I'd like to sync up in person about Alpine Linux oddities before you merge the PR.

File: Dockerfile-slim.template
Comment (jeancochrane) [reply Dockerfile-slim.template]: Sure! Do you think this info should be included in a comment, or in the body of the PR?

File: .travis.yml
Comment (jeancochrane) [reply .travis.yml:5]: Done in https://github.com/azavea/docker-django/pull/66/commits/b01a1df030f164ea0b2993b01828fd8c4f3244cd. Now that we're bumping `gevent` and `psycopg2` to the latest patches, too, there's quite a bit going on in this PR and it's hard to capture it in the title. Do you think we should accept a super long title and update it to include all of the versions that are getting bumped, or is it fine as is?

**Pull request merged**
